### ISSUE TITLE
Floating Base doesn't work with eigen version > 3.2.0

### ISSUE THREAD

**Issue by rlober** [ocra-recipes/ocra-recipes#19]
So I just installed everything on my mac and I have the same problem I was having in ubuntu 16.04 - the fixed base controller works but not the floating base. Really this should be in https://github.com/ocra-recipes/ocra-wbi-plugins but I am not 100% certain that eigen is the cause of this bug. I think it is eigen because the apt-get version of eigen3 on 14.04 is v3.2.0, and on OSX it is v3.2.5 and on 16.04 it is v3.3~beta. My guess is that somewhere in the 5 patches something broke https://github.com/ocra-recipes/eigen_lgsm ... The problem could be more subtle but I will have to test to find out. 

@jeljaik - Can you test on your machine?

First see what version of eigen you have: `brew list eigen`

then open up an `icub.world` in gazebo and run:

```
ocra-icub-server --floatingBase --taskSet wholeBodyTaskSet
```

**Comment by jeljaik** [comment]
Eigen version:

``` bash
/usr/local/Cellar/eigen/3.2.8/include/eigen3/ (365 files)
/usr/local/Cellar/eigen/3.2.8/share/cmake/Modules/FindEigen3.cmake
/usr/local/Cellar/eigen/3.2.8/share/pkgconfig/eigen3.pc
```

Running `ocra-icub-server` with the default initial position of the robot (straight legs) gives me something like this:

<img width="254" alt="screen shot 2016-05-12 at 16 25 00" src="https://cloud.githubusercontent.com/assets/5030039/15217983/2aac962a-185e-11e6-896a-1aafa2d1020f.png">

I have a feeling it's not what we want. Do you get the same?

**Comment by rlober** [comment]
exact same... I am going to push some logging code to dev

**Comment by ahoarau** [comment]
@rlober yep it's probably eigen_lgsm. I'll check 3.3 changeLog.

**Comment by rlober** [comment]
Ok pushed a commit to dev which logs the base pose and twist. @ahoarau, could you run the example for like 2 seconds and send me the logs? They will be wherever you launch the controller from.

**Comment by ahoarau** [comment]
Base Pose : 

```
0.0028796   -0.0681085  0.597488    -1.3782e-05 0.00416414  -2.50146e-05    0.999991
0.00287963  -0.0681092  0.597488    -1.36821e-05    0.00416416  -2.56106e-05    0.999991
0.00287959  -0.0681089  0.597488    -1.37996e-05    0.00416389  -2.55033e-05    0.999991
0.00288597  -0.0681019  0.597488    -4.32393e-05    0.00411658  -2.21757e-05    0.999992
0.00289367  -0.0681034  0.597488    -6.15122e-05    0.00410873  -2.28992e-05    0.999992
0.00290017  -0.0681021  0.597488    -7.18399e-05    0.00411076  -2.17863e-05    0.999992
```

Base twist : 

```
  2.3408e-05 -1.15739e-05  1.67367e-08  -6.3755e-06 -1.41984e-05 -1.56269e-06
 4.86093e-05 -2.54076e-05  -3.4836e-07 -1.40058e-05 -2.93087e-05 -3.24174e-06
 5.89787e-05 -2.85341e-05 -2.79961e-07  -1.5312e-05 -3.52771e-05 -3.94111e-06
  5.0978e-05 -0.000237587  0.000128898 -1.71318e-06 -2.32035e-05 -3.29575e-06
 4.40883e-05 -0.000452336  0.000321787  2.77869e-05 -1.38796e-05 -2.72948e-06
 2.99003e-05 -0.000623064  0.000531122  6.81824e-05 -9.48573e-07 -1.70864e-06
 3.88912e-05 -0.000747713  0.000728769  0.000111919 -2.42583e-06 -2.30236e-06
 4.14209e-05 -0.000814521  0.000902216  0.000164804 -4.62136e-07 -2.53516e-06
```

**Comment by rlober** [comment]
Could you attach the logs?

**Comment by ahoarau** [comment]
[base_pose_log.txt](https://github.com/ocra-recipes/ocra-recipes/files/261808/base_pose_log.txt)
[base_twist_log.txt](https://github.com/ocra-recipes/ocra-recipes/files/261807/base_twist_log.txt)

**Comment by rlober** [comment]
Yay ! :cake:

**Comment by rlober** [comment]
Good results are in green - falling is in blue.

**Base Pose**
![base_poses](https://cloud.githubusercontent.com/assets/3417094/15222741/6a8fac88-1871-11e6-95a8-9099c2a8ed04.png)

**Base Twist**
![base_twists](https://cloud.githubusercontent.com/assets/3417094/15222742/6a90c8de-1871-11e6-8e4c-6ffe2742b584.png)

**Comment by rlober** [comment]
Here is a zoom in on the first 20 timesteps (0.2 s)

**Pose**
![base_poses](https://cloud.githubusercontent.com/assets/3417094/15239325/eddf2c56-18e1-11e6-9997-8958e8fc1e0e.png)

**Twist**
![base_twists](https://cloud.githubusercontent.com/assets/3417094/15239326/eddf4df8-18e1-11e6-8f2a-5c5451254193.png)

**Comment by ahoarau** [comment]
it seems to be correctly estimated during the first tests, could you also record the torque commands we're sending to the robot ?

**Comment by rlober** [comment]
[torques_log.txt](https://github.com/ocra-recipes/ocra-recipes/files/262677/torques_log.txt)

**Comment by rlober** [comment]
![torques](https://cloud.githubusercontent.com/assets/3417094/15240845/a2d75a34-18ec-11e6-9500-529ea0553a90.png)

**Comment by rlober** [comment]
Ok so findings so far:
1. Robot falls in floating base (both wocra and hocra quadprog or qpoases)
2. Robot works fine in fixed base
3. This means that there is something wrong when the model dynamics are floating base.
4. In [`ocra::Model`](https://github.com/ocra-recipes/ocra-recipes/blob/master/ocra/src/control/Model.cpp#L83) `setFreeFlyerPosition` and `setFreeFlyerVelocity` are called when using floating base. 

``` cpp
void Model::setFreeFlyerPosition(const Eigen::Displacementd& H_root)
  {
    assert(!_fixedRoot);
    doSetFreeFlyerPosition(H_root);
    //we do not set H_root in q_root on purpose: H_root is in SE3 while q_root is in R^6
    propagate<EVT_CHANGE_VALUE>();
  }

  void Model::setFreeFlyerVelocity(const Eigen::Twistd& T_root)
  {
    assert(!_fixedRoot);
    doSetFreeFlyerVelocity(T_root);
    getRootVelocityVariable().setValue(T_root.get());
    propagate<EVT_CHANGE_VALUE>();
  }
```
1. `doSetFreeFlyerPosition` and  `doSetFreeFlyerVelocity` are called in ocra-wbi-plugins by [`OcraWbiModel`](https://github.com/ocra-recipes/ocra-wbi-plugins/blob/dev/ocra-icub/src/OcraWbiModel.cpp#L640). These basically just set values in the pimpl which are used for floating base calcs of mass mat, nonlin terms, and grav terms. I have not thuroughly checked these.
2. The floating base estimations seem fine, but the calculated torques are far too high.
3. Somehow I feel that something is happening in `propagate<EVT_CHANGE_VALUE>();`

**Comment by rlober** [comment]
Ok so after 3 days of debugging, and feeling like this:
[![image](https://cloud.githubusercontent.com/assets/3417094/15427171/6e04864a-1e93-11e6-985a-d37e8b0e8091.png)](https://youtu.be/dsx2vdn7gpY?start=0&end=13&version=3)

...the issue has been fixed by f85f39b53618e346d83ff00fb5003c184b3d3928

Basically the reason the controller was calculating erroneous torques was because the contact set offsets were not being parsed correctly by the XML task parser. This resulted in 4 contact points at the same locations on the left and right foot. As a result, the inequality constraint was no longer full rank and the QP solves were screwed up... To put things technically.  

On machines using gcc4 the parser was working fine. When compiled with gcc5 the `const char *` passed from `getDisplacementArgs(TiXmlElement* xmlElem)` to `stringToVectorXd(const char * valueString)` was empty and therefore all of the contact offsets were empty. I am still not 100% sure as to why the `const char *` was not being passed correctly but I guess it is related to this:

https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html

> In the GCC 5.1 release libstdc++ introduced a new library ABI that includes new implementations of std::string and std::list. These changes were necessary to conform to the 2011 C++ standard which forbids Copy-On-Write strings and requires lists to keep track of their size.

Thanks you to @traversaro for pointing this out in a related issue: https://github.com/robotology/idyntree/issues/163#issuecomment-220271816

Master was working because it still used `taskSequences` where no parsing occurs, but when we tested it with `taskSets` which are parsed, we saw the same behavior.

I am closing both issues but I just want to thank @jeljaik and @ahoarau for all their hard work, because it was long... and hard...
